File: cpp/problem292.cpp
#include <iostream>
#include <vector>
using namespace std;

// Class: Solution with mathematica thoughts.
// Description:
//  - from mathematica, if n % 4 == 0, then i will fail!
//  - constant space and time costs.
// Creat: Yangyao, date 20181004
// References: none.
class Solution {
public:
    bool canWinNim(int n) {
        return n%4;
    }
};

// Class: Solution with iteration.
// Description:
//  - use a deque-like structure to keep former result.
//  - constant space cost.
//  - linear time cost = Theta(n).
// Creat: Yangyao, date 20181004
// References: none.
class Solution2 {
public:
    bool canWinNim(int n) {
        bool history[3] = {true, true, false};
        int findex = 2;
        for (int i = 0; i < n; ++i){
            findex = (findex+1)%3;
            history[findex] = (history[0] && history[1] && history[2])?false:true;
        }
        return history[findex];
    }
};

template<typename T>
void check_time( T &sol, int repeat = 1000, int n = 100 ){
    clock_t t1 = clock();
    for (int i = 0; i < repeat; ++i){
        sol.canWinNim(n);
    }
    clock_t t2 = clock();
    cout << "time cost: " << double(t2-t1)/CLOCKS_PER_SEC << endl;
}


// Func: Driver function to test solutions
int main(int argc, char const *argv[]){
    Solution sol;
    Solution2 sol2;

    // Test solutions by different inputs.
    cout << "\tsolution and solution2\n";
    for (int i = 1; i <= 10; ++i){
        cout << i << '\t' << sol.canWinNim(i) << '\t' << sol2.canWinNim(i) << '\n';
    }

    // Compare efficiency of solutions.
    check_time(sol);
    check_time(sol2);

    return 0;
}

File: swift/problem292.swift
import Dispatch

// recursive
class Solution {
    func canWinNim(_ n: Int) -> Bool {
        if (n>3) {
            if (canWinNim(n-1) && canWinNim(n-2) && canWinNim(n-3)) {
                return false
            } else {
                return true
            }
        } else {
            return true
        }
    }
}

//iteration
class Solution2 {
    func canWinNim(_ n: Int) -> Bool {
        if (n<4) {
            return true
        } else {
            var history = [true,true,true]
            var index_i = 2
            for _ in 4...n {
                index_i = (index_i+1)%3
                history[index_i] = !(history[0] && history[1] && history[2])
            }
            return history[index_i]
        }
    }
}

//math
class Solution3 {
    func canWinNim(_ n : Int) -> Bool {
        return (n%4 != 0)
    }
}

//test time consuming for different methods
var test1 = Solution()
var test2 = Solution2()
var test3 = Solution3()

//recursive method is extremely slow
var time_start = DispatchTime.now().uptimeNanoseconds
for i in 1...40 {
    _ = test1.canWinNim(i)
}
var time_end = DispatchTime.now().uptimeNanoseconds
print("recursive",(time_end-time_start)/1000000,"ms")

time_start = DispatchTime.now().uptimeNanoseconds
for i in 1...32768 {
    _ = test2.canWinNim(i)
}
time_end = DispatchTime.now().uptimeNanoseconds
print("iteration",(time_end-time_start)/1000000,"ms")

time_start = DispatchTime.now().uptimeNanoseconds
for i in 1...32768 {
    _ = test3.canWinNim(i)
}
time_end = DispatchTime.now().uptimeNanoseconds
print("math",(time_end-time_start)/1000000,"ms")
